src/aliceVision/matchingImageCollection/GeometricFilterMatrix_HGrowing.hpp

```diff
@@ -0,0 +1,565 @@
+          const SIOPointFeature & fJ  = siofeatures_J.at(match._j);
+          svgStream->drawCircle(fI.x(), fI.y(), 1, svg::svgStyle().stroke("white",2.0));
+          svgStream->drawCircle(fJ.x() + viewI.getWidth(), fJ.y(), 1, svg::svgStyle().stroke("white", 2.0));
+        }
+      }
+      
+      for (IndexT iH = 0; iH < _maxNbHomographies; ++iH)
+      {
+        std::set<IndexT> usedMatchesId, bestMatchesId;
+        Mat3 bestHomographie;
+        
+        // -- Estimate H using homogeaphy-growing approach:
+        
+#pragma omp parallel for // (: huge optimization but modify results a little)
+        for (int iMatch = 0; iMatch < remainingMatches.size(); ++iMatch)
+        {
+          // Growing a homography from one match ([F.Srajer, 2016] algo. 1, p. 20)  
+          // each match is used once only per homography estimation (increases computation time) [1st improvement ([F.Srajer, 2016] p. 20) ] 
+          if (usedMatchesId.find(iMatch) != usedMatchesId.end()) 
+            continue;
+          
+          std::set<IndexT> planarMatchesId; // be careful: it contains the id. in the 'remainingMatches' vector not 'putativeMatches' vector.
+          Mat3 homographie;
+          
+          if(!growHomography(siofeatures_I, siofeatures_J, remainingMatches, iMatch, planarMatchesId, homographie) == EXIT_SUCCESS)
+            continue;
+          
+#pragma omp critical
+          usedMatchesId.insert(planarMatchesId.begin(), planarMatchesId.end());
+          
+          if (planarMatchesId.size() > bestMatchesId.size())
+          {
+#pragma omp critical
+            {
+              bestMatchesId = planarMatchesId; // be careful: it contains the id. in the 'remainingMatches' vector not 'putativeMatches' vector.
+              bestHomographie = homographie;
+            }
+          }
+        } // 'iMatch'
+        
+        // -- Refine H using Ceres minimizer:
+        {
+          ceres::Problem problem;
+          
+          for(IndexT matchId : bestMatchesId)
+          {
+            IndMatch match = remainingMatches.at(matchId);
+            
+            Vec2 x1 = siofeatures_I.at(match._i).coords().cast<double>();
+            Vec2 x2 = siofeatures_J.at(match._j).coords().cast<double>();
+            
+            RefineHRobustCostFunctor 
+                *costFun = 
+                new RefineHRobustCostFunctor(x1, x2, _homographyTolerance);
+            
+            problem.AddResidualBlock(
+                  new ceres::AutoDiffCostFunction<
+                  RefineHRobustCostFunctor,
+                  1,
+                  9>(costFun), 
+                  NULL, 
+                  bestHomographie.data());
+          }
+          
+          ceres::Solver::Options solverOpt;
+          solverOpt.max_num_iterations = 10;
+          
+          ceres::Solver::Summary summary;
+          ceres::Solve(solverOpt,&problem,&summary);
+          
+          bestHomographie /= bestHomographie(2,2);
```

Comment:
it shuld check that it is not equal to zero to avoid zero division

Reply:
Not sure what to do in that case and if we really need.
Currently, it will generate NaN values.

Reply:
Yes but that's an error case and we should stop, isn't it?

Reply:
no it's not necessarily an error, it could happen that the 9th element is (close to) 0